### ISSUE TITLE
BUG: Server Crashes with Assertion Error `GeneralUtils::ASCIIToUTF16` Assertion `c > 0 && c <= 127`

### ISSUE THREAD

**Issue by TheMatt2** [DarkflameUniverse/DarkflameServer#350]
### Make sure you've done the following:

- [X] I have checked that this issue has not already been reported.

- [X] I have validated that this issue is not a syntax error of either MySQL or SQLite.

- [X] I have pulled the latest version of the main branch of DarkflameServer and have confirmed that the issue exists there.

### Issue Description

The server will crash if `GeneralUtils::ASCIIToUTF16`  is passed a string that falls outside of this range.
The server does not actually exit, but hangs. Any player trying to login will be stuck on 0% in world load.

Specifically the server seems to be crashing on processing the text `8066 - Model Reward Café Table`.
The character "é" seems to evaluate to the ascii value 233, which falls outside of the assertion made by `GeneralUtils::ASCIIToUTF16` that all values are less than 127.

### Reproduction steps

A user with appropriate permission can type into the char window:

    /lookup table

The server will crash a few seconds later.

### Expected Behavior

Accented characters should be loaded correctly from ASCII to UTF16.
On assertion failure, the server should cleanly exit.

### Environment

Ubuntu under Docker

Traceback
```
darkflameserverdev     | WorldServer: /DarkflameServer/dCommon/GeneralUtils.cpp:53: std::u16string GeneralUtils::ASCIIToUTF16(const string&, size_t): Assertion `c > 0 && c <= 127' failed.
darkflameserverdev     | Fatal error 6
darkflameserverdev     | Stacktrace:
darkflameserverdev     | ./WorldServer(+0xe739f)[0x560b8ae5939f]
darkflameserverdev     | /lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0)[0x7f8f35a1b3c0]
darkflameserverdev     | /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb)[0x7f8f3550518b]
darkflameserverdev     | /lib/x86_64-linux-gnu/libc.so.6(abort+0x12b)[0x7f8f354e4859]
darkflameserverdev     | /lib/x86_64-linux-gnu/libc.so.6(+0x25729)[0x7f8f354e4729]
darkflameserverdev     | /lib/x86_64-linux-gnu/libc.so.6(+0x36f36)[0x7f8f354f5f36]
darkflameserverdev     | ./WorldServer(+0xe84ae)[0x560b8ae5a4ae]
darkflameserverdev     | ./WorldServer(+0x2aeff8)[0x560b8b020ff8]
darkflameserverdev     | ./WorldServer(+0x2638d0)[0x560b8afd58d0]
darkflameserverdev     | ./WorldServer(+0x246b65)[0x560b8afb8b65]
darkflameserverdev     | [03-01-22 21:04:07] [WorldServer] Attempting to shutdown world, zone (1100), instance (2), max 10 seconds...
darkflameserverdev     | [03-01-22 21:04:18] [ChatServer] A server has disconnected, erasing their connected players from the list.
darkflameserverdev     | [03-01-22 21:04:21] [MasterServer] A server has lost the connection
```

**Comment by TheMatt2** [comment]
Relevant code

https://github.com/DarkflameUniverse/DarkflameServer/blob/8ed7690b966fc53a9ea0fd221a1275735d429ab5/dCommon/GeneralUtils.cpp#L53

**Comment by TheMatt2** [comment]
Perhaps remove assertion?

**Comment by Xiphoseer** [comment]
ASCII is only defined to be 7 bit - that's on purpose, so the issue is passing locale / database strings which are latin1 through it in the first place because those actually need a special lookup table to be transformed to unicode strings.

If we don't have such a function already, we may need to add it and use it wherever appropriate.

**Comment by TheMatt2** [comment]
There is also the issue that it seems the client isn't properly displaying UTF32 values.

This is from a test build that has the assertion commented out.
So what value should this be encoded to?

![image](https://user-images.githubusercontent.com/22862051/148139977-eaf10d59-da72-43b9-a9e0-503609e3b19d.png)

**Comment by Xiphoseer** [comment]
So I would argue this line is the culprit:

https://github.com/DarkflameUniverse/DarkflameServer/blob/a9860388b2b34d4a52c63c673d0e4a94ec089ebc/dGame/dUtilities/SlashCommandHandler.cpp#L1190

Of course a valid fix may be to always treat 8bit strings as latin1, but I don't think that's better. The encoding of sqlite strings (which are used in /lookup afaict) is always utf8 (or utf16), so it might be better to use that conversion here.

**Comment by TheMatt2** [comment]
Just as a sanity check, the text does render correctly in the client when the item is viewed.
![image](https://user-images.githubusercontent.com/22862051/148140379-6ff55c75-fbb4-43fe-88bd-202789a9fb66.png)

Though it seems `ChatPackets::SendSystemMessage()` requires utf16.
How is the client expecting the text to be utf8?

**Comment by Xiphoseer** [comment]
The client loads its text from locale.xml, that's entirely separate. And yes, the chat messages are *windows* wstrings which is technically UCS-2 in some places but for practical purposes equivalent to UTF-16 i.e. `std::u16string`

**Comment by Xiphoseer** [comment]
I just checked our internal version history and it looks like a utility for utf8 -> was never implemented. Maybe that's because that is in c++ by default using `std::codecvt_utf8_utf16`.

c.f. https://stackoverflow.com/questions/7232710/convert-between-string-u16string-u32string

**Comment by Xiphoseer** [comment]
> There is also the issue that it seems the client isn't properly displaying UTF32 values.
> 
> This is from a test build that has the assertion commented out.
> So what value should this be encoded to?
> 
> ![image](https://user-images.githubusercontent.com/22862051/148139977-eaf10d59-da72-43b9-a9e0-503609e3b19d.png)
> 

Yeah, that's exactly what two utf8 bytes (making one é codepoint) treated as two separate utf16 codepoints might look like.

**Comment by Xiphoseer** [comment]
That would mean the LU font doesn't have glyphs for https://codepoints.net/U+00D0 and https://codepoints.net/U+00A0, which sounds not unlikely

**Comment by Xiphoseer** [comment]
TLDR: The correct fix is a proper conversion helper function UTF8toUTF16 using std::codecvt_utf8_utf16 as the implementation

**Comment by Jettford** [comment]
This should probably be low priority? This only occurs if a GM sends a character that regular players can't on a server build in debug. (as long as I understand the issue here)

**Comment by Xiphoseer** [comment]
> This should probably be low priority? This only occurs if a GM sends a character that regular players can't on a server built in debug. (as long as I understand the issue here) 

IMHO the underlying issue is likely triggerable elsewhere too. ASCIIToUTF8 is used in at least a few other cases where they shouldn't, like property names.

When I added it, it was for animation keys, which aren't user input.

**Comment by TheMatt2** [comment]
I suspect other player interactions with the item `Model Reward Café Table` may cause this as well.
This isn't the GM sending a character regular player's can't. Rather, its what happens when this particular model is interacted with in the code. The `/lookup` command is just a handy way to make the server interact with lots of model names.

Furthermore, if some other method were to cause this server crash, it would be very difficult to reproduce, as the server would crash before ever showing `Model Reward Café Table` to the user.

**Comment by Xiphoseer** [comment]
> I suspect other player interactions with the item `Model Reward Café Table` may cause this as well.
> This isn't the GM sending a character regular player's can't. Rather, its what happens when this particular model is interacted with in the code. The `/lookup` command is just a handy way to make the server interact with lots of model names.
> 
> Furthermore, if some other method were to cause this server crash, it would be very difficult to reproduce, as the server would crash before ever showing `Model Reward Café Table` to the user.

That's unlikely. The client is perfectly capable of handling objects with ° or é in their names -- they always use wstring. What happend here is that a *server-side* command takes data from a DB technology the game never used (sqlite), with a different encoding (utf8) and sends it to chat with a conversion function that wasn't intended for that.

It's data from the outside, even if originally from the game that's the issue.

**Comment by Xiphoseer** [comment]
Closing this because #658 should make it not crash anymore at least.
